src/merkle/mmr/partial.rs

```diff
@@ -0,0 +1,382 @@
+            "if there is an odd element, a merge is required"
+        );
+
+        // keeps track of how many data elements from the update have been consumed
+        let mut update_count = 0;
+
+        if merges != 0 {
+            // starts at the smallest peak and follows the merged peaks
+            let mut peak_idx = forest_to_root_index(self.forest);
+
+            // match order of the update data while applying it
+            self.peaks.reverse();
+
+            // set to true when the data is needed for authentication paths updates
+            let mut track = self.track_latest;
+            self.track_latest = false;
+
+            let mut peak_count = 0;
+            let mut target = 1 << merges.trailing_zeros();
+            let mut new = updates[0];
+            update_count += 1;
+
+            while target < largest {
+                // check if either the left or right subtrees have saved for authentication paths.
+                // If so, turn tracking on to update those paths.
+                if target != 1 && !track {
+                    let left_child = peak_idx.left_child();
+                    let right_child = peak_idx.right_child();
+                    track = self.nodes.contains_key(&left_child)
+                        | self.nodes.contains_key(&right_child);
+                }
+
+                // update data only contains the nodes from the right subtrees, left nodes are
+                // either previously known peaks or computed values
+                let (left, right) = if target & merges != 0 {
+                    let peak = self.peaks[peak_count];
+                    peak_count += 1;
+                    (peak, new)
+                } else {
+                    let update = updates[update_count];
+                    update_count += 1;
+                    (new, update)
+                };
+
+                if track {
+                    self.nodes.insert(peak_idx.sibling(), right);
+                }
+
+                peak_idx = peak_idx.parent();
+                new = Rpo256::merge(&[left, right]);
+                target <<= 1;
+            }
+
+            debug_assert!(peak_count == (merges.count_ones() as usize));
+
+            // restore the peaks order
+            self.peaks.reverse();
+            // remove the merged peaks
+            self.peaks.truncate(self.peaks.len() - peak_count);
+            // add the newly computed peak, the result of the merges
+            self.peaks.push(new);
+        }
+
+        // The rest of the update data is composed of peaks. None of these elements can contain
+        // tracked elements because the peaks were unknown, and it is not possible to add elements
+        // for tacking without authenticating it to a peak.
+        self.peaks.extend_from_slice(&updates[update_count..]);
+        self.forest = forest;
+
+        debug_assert!(self.peaks.len() == (self.forest.count_ones() as usize));
```

Comment:
Humm, at this point it is already a programming error. We have modified the `PartialMmr` and if this assert fails we have a broken invariant.

But that is a good point, it would be an attack vector for a rogue node. I'm adding some validation in addition to this.

Reply:
I think that's fine. We'll have a separate check in the client to make sure that hash of the peaks matches the `chain_root`.